torch/nn/utils/_per_sample_grad.py

```diff
@@ -1,29 +1,30 @@
+import functools
+
 import torch
 from torch.nn.utils._stateless import functional_call
 from torch.nn.utils._expanded_weights.expanded_weights_impl import ExpandedWeight
 
 # dependency on `functional_call` means that this can't be exposed in utils
 # without creating circular dependency
-def call_for_per_sample_grads(module, batch_size, args, kwargs=None):
+def call_for_per_sample_grads(module, batch_size, loss_reduction="mean"):
```

Comment:
I don't really like having a loss_reduction argument because the loss is not guaranteed to have a loss_reduction argument (in general) but also because nll_loss's mean argument is different from other losses.

[(docs)](https://pytorch.org/docs/stable/generated/torch.nn.NLLLoss.html#torch.nn.NLLLoss) nll_loss's mean argument doesn't mean "divide by batch_size", it means divide by the sum of the relevant weights. So if someone uses nll_loss they won't actually be able to reproduce the semantics of a for-loop.


To check my understanding about the motivation, a forward+backward pass looks like the following:
```
model = ...
targets = ...
criterion = CrossEntropyLoss(reduction='mean')
result = call_for_per_sample_grads(model, batch_size, input)
loss = criterion(result, targets)
loss.backward()
```
what happens if you move the criterion call to inside the model? Does that give the desired output?

Reply:
I see, thanks for the detailed explanation. This seems like a fundamental limitation of the current design; there is no real way to detect that a user is doing "batch-dependent operations" (in this case we are talking about dividing the input by 1/batch_size, but another example is batch_norm) so we resolve this via user intervention by asking them to provide us a string

Reply:
Resolved offline: we're just going to default loss reduction to 'sum'. There's a "long term expanded weights" design to fix this, but not much we can do in the short term.